src/PantheonShell.vala

```diff
@@ -0,0 +1,46 @@
+public class Gala.PantheonShell : Object {
+    public static void init (Meta.Display display) requires (instance == null) {
+        instance = new PantheonShell (display);
+    }
+
+    private static PantheonShell? instance = null;
+
+    public Meta.Display display { get; construct; }
+
+    private Meta.WaylandClient dock_client;
+
+    private PantheonShell (Meta.Display display) {
+        Object (display: display);
+    }
+
+    construct {
+        display.window_created.connect ((window) => {
+            if (dock_client != null && dock_client.owns_window (window)) {
+                setup_dock_window (window);
+            }
+        });
+
+        var subprocess_launcher = new GLib.SubprocessLauncher (NONE);
+        try {
+            dock_client = new Meta.WaylandClient (subprocess_launcher);
+            string[] args = {"io.elementary.dock"};
```

Comment:
For a more useful way to test this replace dock with wingpanel and remove the TryExec for io.elementary.wingpanel from the wayland session settings and it works really fine apart from maximized windows disappearing behind the panel :)